janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphBlueprintsGraph.java

```diff
@@ -326,6 +327,7 @@ public void close() {
         void close(Transaction tx) {
             closeConsumerInternal.get().accept(tx);
```

Comment:
Is it OK that the same operation will be executed during `super.close();` execution?
I.e. : 
```
((Consumer)this.closeConsumerInternal.get()).accept(this);
this.closeConsumerInternal.remove();
this.readWriteConsumerInternal.remove();
```

Reply:
From one hand is it no, but this approach bugs free [CLOSE_BEHAVIOR](https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/Transaction.java#L165) operations may be invoked twise

Also, I think we may remove `GraphTransaction.close(..) and GraphTransaction.close()` because base `AbstractTransaction.close` do same things. 

And in `JanusGraphBlueprintsTransaction` direct invoke close, like this. AbstractThreadedTransaction should not handle onReadeWrite/onClose operations, see docs on methods

```
    @Override
            public void doClose() {
                getGraph().tinkerpopTxContainer.close();

                // calling super will clear listeners
                super.doClose();
            }
```

Reply:
@mad I also think that we need to remove `GraphTransaction.close()` but I am not sure about `GraphTransaction.close(Transaction tx)`. The code there a little bit confusing for me that is why I don't know if we should remove this method or not.
The method `GraphTransaction.close(Transaction tx)` executes `closeConsumerInternal.get().accept(tx);` and the method `AbstractThreadLocalTransaction.close()` executes `closeConsumerInternal.get().accept(this);`. I.e. the difference in `this` and `tx`.
Currently in `JanusGraphBlueprintsTransaction.doClose()` we execute `getGraph().tinkerpopTxContainer.close(this);` which is `GraphTransaction.close(Transaction tx)` which invokes `closeConsumerInternal.get().accept(tx);`. If we change the code in JanusGraphBlueprintsTransaction to your posted code then we will execute `closeConsumerInternal.get().accept(this);` instead of `closeConsumerInternal.get().accept(tx);`. 

It is a little bit confusing for me and I don't know how should it be. It is clear for me that we should remove `GraphTransaction.close()` but as I said I don't know about `GraphTransaction.close(Transaction tx)`. Maybe I am missing something. Can you explain please why `getGraph().tinkerpopTxContainer.close(this);` and `getGraph().tinkerpopTxContainer.close();` will be the same if invoked in `JanusGraphBlueprintsTransaction`?

Reply:
@porunov I think we do not need do `getGraph().tinkerpopTxContainer.close();` because when we have `StandardJanusGraphTx` instance we should commit/rollback manually -- this is manual mode, see javadoc for `JanusGraph.newTransaction` and if tx not commit/rolled backed we need throw an exception like in `CLOSE_BEHAVIOR.MANUAL` mode . When we work with `GraphTransaction` then `AbstractThreadLocalTransaction.doClose` do all auto work when tx is closed